### PR TITLE
Implement components needed for packet clearing

### PR DESCRIPTION
**Pull request by romac**
Closes: #160

## Description

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by romac** [comment]
Next step is to implement https://github.com/informalsystems/hermes/blob/master/crates/relayer/src/chain/counterparty.rs?plain=1#L581 as a generic component.

**Comment by ljoss17** [comment]
> Next step is to implement https://github.com/informalsystems/hermes/blob/master/crates/relayer/src/chain/counterparty.rs?plain=1#L581 as a generic component.

The commit 3f9177e2e8594d20eb1740d539208685edd8918d follows the implementation to clear pending acks for the Cosmos component.
The commit f80df3bce0970392ac5f6977fdd8c830756300e1 wires create a new component to clear both recv and acks.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    pub packet: Relay::Packet,
+}
+
+#[async_trait]
+impl<Relay> Task for RelayPacketTask<Relay>
+where
+    Relay: CanRelayPacket,
+{
+    async fn run(self) {
+        let _ = self.relay.relay_packet(&self.packet).await;
```
Comment (romac) [inline crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:30]: We should probably deal with the error that is potentially returned here, eg. log it.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    Relay::Runtime: CanRunConcurrentTasks,
+{
+    async fn clear_packets(
+        relay: &Relay,
+        src_channel_id: &ChannelIdOf<Relay::SrcChain, Relay::DstChain>,
+        src_port_id: &PortIdOf<Relay::SrcChain, Relay::DstChain>,
+        dst_channel_id: &ChannelIdOf<Relay::DstChain, Relay::SrcChain>,
+        dst_port_id: &PortIdOf<Relay::DstChain, Relay::SrcChain>,
+    ) -> Result<(), Relay::Error> {
+        ClearReceivePackets::clear_packets(
```
Comment (romac) [inline crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs:56]: @soareschen Did I understand correctly earlier that you were suggesting that we run these two tasks concurrently?

File: tools/integration-test/src/tests/packet_clear.rs [hunk 3]
```diff
@@ -238,3 +248,205 @@ impl BinaryChannelTest for IbcClearPacketTest {
         Ok(())
     }
 }
+
+pub struct IbcClearAckTest;
```
Comment (romac) [inline tools/integration-test/src/tests/packet_clear.rs:252]: Awesome!

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs (hunk 1 above)
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:30]: Do you mean something like this?
```suggestion
        let result = self.relay.relay_packet(&self.packet).await;
        if let Err(e) = result {
            let error_desc = format!("{e:#?}");
            self.relay.log_error(&error_desc);
        }
```

Comment (romac) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:30]: Yep! Perhaps with a message indicating we failed to relay.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    pub packet: Relay::Packet,
+}
+
+#[async_trait]
+impl<Relay> Task for RelayPacketTask<Relay>
+where
+    Relay: CanRelayPacket,
+{
+    async fn run(self) {
+        let _ = self.relay.relay_packet(&self.packet).await;
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs]: Done in this commit 67e939233239e3b04b2b0adafebc6f8d567240ea

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs (hunk 1 above)
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs]: I think we would need to add the constraint `HasRelayPacketFields` which might not be the case for non-Cosmos packets.
cc @soareschen

Comment (romac) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs]: Those fields are defined in the [Packet interface in ICS04](https://github.com/cosmos/ibc/blob/main/spec/core/ics-004-channel-and-packet-semantics/README.md) so all IBC-enabled chains should have them IIUC, no?

Comment (soareschen) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs]: I think you only need `Relay::Packet: Debug` or `Display` for formatting the packet.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs (hunk 2 above)
Comment (soareschen) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs:62]: Yup. We can implement this much simpler as follows:

```rust
use alloc::vec;

use crate::chain::traits::types::ibc::HasIbcChainTypes;
use crate::chain::types::aliases::{ChannelIdOf, PortIdOf};
use crate::relay::traits::chains::HasRelayChains;
use crate::relay::traits::packet_clearer::PacketClearer;
use crate::runtime::traits::runtime::HasRuntime;
use crate::runtime::traits::task::{CanRunConcurrentTasks, Task};

use super::ack::ClearAckPackets;
use super::receive_packet::ClearReceivePackets;

pub struct ClearAllPackets;

pub enum ClearOption {
    Receive,
    Ack,
}

pub struct RunPacketClearer<Relay>
where
    Relay: HasRelayChains,
{
    pub relay: Relay,
    pub src_channel_id: ChannelIdOf<Relay::SrcChain, Relay::DstChain>,
    pub src_port_id: PortIdOf<Relay::SrcChain, Relay::DstChain>,
    pub dst_channel_id: ChannelIdOf<Relay::DstChain, Relay::SrcChain>,
    pub dst_port_id: PortIdOf<Relay::DstChain, Relay::SrcChain>,
    pub clear_option: ClearOption,
}

impl<Relay> Task for RunPacketClearer<Relay>
where
    Relay: HasRelayChains,
    ClearReceivePackets: PacketClearer<Relay>,
    ClearAckPackets: PacketClearer<Relay>,
{
    async fn run(self) {
        let _ = match self.clear_option {
            ClearOption::Receive => {
                ClearReceivePackets::clear_packets(
                    &self.relay,
                    &self.src_channel_id,
                    &self.src_port_id,
                    &self.dst_channel_id,
                    &self.dst_port_id,
                )
                .await
            }
            ClearOption::Ack => {
                ClearAckPackets::clear_packets(
                    &self.relay,
                    &self.src_channel_id,
                    &self.src_port_id,
                    &self.dst_channel_id,
                    &self.dst_port_id,
                )
                .await
            }
        };
    }
}

impl<Relay, SrcChain, DstChain> PacketClearer<Relay> for ClearAllPackets
where
    Relay: Clone + HasRuntime + HasRelayChains<SrcChain = SrcChain, DstChain = DstChain>,
    SrcChain: HasIbcChainTypes<DstChain>,
    DstChain: HasIbcChainTypes<SrcChain>,
    Relay::Runtime: CanRunConcurrentTasks,
    ClearReceivePackets: PacketClearer<Relay>,
    ClearAckPackets: PacketClearer<Relay>,
    SrcChain::ChannelId: Clone,
    SrcChain::PortId: Clone,
    DstChain::ChannelId: Clone,
    DstChain::PortId: Clone,
{
    async fn clear_packets(
        relay: &Relay,
        src_channel_id: &SrcChain::ChannelId,
        src_port_id: &SrcChain::PortId,
        dst_channel_id: &DstChain::ChannelId,
        dst_port_id: &DstChain::PortId,
    ) -> Result<(), Relay::Error> {
        let receive_packet_task = RunPacketClearer {
            relay: relay.clone(),
            src_channel_id: src_channel_id.clone(),
            src_port_id: src_port_id.clone(),
            dst_channel_id: dst_channel_id.clone(),
            dst_port_id: dst_port_id.clone(),
            clear_option: ClearOption::Receive,
        };

        let ack_packet_task = RunPacketClearer {
            relay: relay.clone(),
            src_channel_id: src_channel_id.clone(),
            src_port_id: src_port_id.clone(),
            dst_channel_id: dst_channel_id.clone(),
            dst_port_id: dst_port_id.clone(),
            clear_option: ClearOption::Ack,
        };

        relay
            .runtime()
            .run_concurrent_tasks(vec![receive_packet_task, ack_packet_task])
            .await;

        Ok(())
    }
}
```

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs (hunk 1 above)
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs]: done in faa05f1536a7af583e424b55c81f00a24510cfe1

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs]: done in faa05f1536a7af583e424b55c81f00a24510cfe1

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs [hunk 5]
```diff
@@ -19,62 +12,97 @@ use super::receive_packet::ClearReceivePackets;
...
+                ClearAckPackets::clear_packets(
+                    &self.relay,
+                    &self.src_channel_id,
+                    &self.src_port_id,
+                    &self.dst_channel_id,
+                    &self.dst_port_id,
+                )
+                .await
+            }
+        };
```
[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs:60]: Oh right, nice catch

[the remark this answers is not in the document]
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs:62]: done in 7b4a8b64e446b54cbbff857b1ba9c5a2b1d7c0ca

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+#[async_trait]
+impl<Relay> PacketClearer<Relay> for ClearAckPackets
+where
+    Relay: Clone + CanRelayPacket + HasRuntime + CanRaiseRelayChainErrors + CanLog,
+    Relay::DstChain: CanQueryUnreceivedAcksSequences<Relay::SrcChain>
+        + CanQueryPacketAcknowledgements<Relay::SrcChain>,
+    Relay::SrcChain:
+        CanQueryPacketCommitments<Relay::DstChain> + CanQuerySendPackets<Relay::DstChain>,
+    Relay::Runtime: CanRunConcurrentTasks,
+    Relay::Packet: Display,
```
Comment (soareschen) [inline crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:52]: I think the side here is mistaken. When clearing ack packets, we are querying the unreceived acks from the source chain. Remember that in a relay context, it is the destination chain that sends the ack, and the source chain that receives the ack.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+#[async_trait]
+impl<Relay> Task for RelayPacketTask<Relay>
+where
+    Relay: CanRelayPacket + CanLog,
+    Relay::Packet: Display,
+{
+    async fn run(self) {
+        if let Err(e) = self.relay.relay_packet(&self.packet).await {
```
Comment (soareschen) [inline crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:34]: To clear an ack packet, we only need `Relay: CanRelayAckPacket`.

**Review by soareschen — CHANGES_REQUESTED** [review]

File: crates/relayer/relayer-components/src/chain/traits/queries/unreceived_acks_sequences.rs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    ///
+    /// For example, given a request with the sequence numbers `[5,6,7,8]`, a
+    /// response of `[7,8]` would indicate that packets 5 & 6 were acknowledged,
+    /// while packets 7, 8 were not.
+    async fn query_unreceived_acknowledgments_sequences(
+        &self,
+        channel_id: &Self::ChannelId,
+        port_id: &Self::PortId,
+        packet_ack_sequences: &[Counterparty::Sequence],
+    ) -> Result<Vec<Counterparty::Sequence>, Self::Error>;
```
Comment (soareschen) [inline crates/relayer/relayer-components/src/chain/traits/queries/unreceived_acks_sequences.rs:27]: The sequences here should probably be `Self::Sequence`.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            .await
+            .map_err(Relay::raise_error)?;
+
+        let acks_and_height_on_counterparty = dst_chain
+            .query_packet_acknowlegements(dst_channel_id, dst_port_id, &commitment_sequences)
+            .await
+            .unwrap();
+
+        if let Some((acks_on_counterparty, height)) = acks_and_height_on_counterparty {
+            let unreceived_ack_sequences = dst_chain
```
Comment (soareschen) [inline crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:75]: This should probably be queried from the source chain.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs (hunk 6 above)
Comment (ljoss17) [reply crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:52]: Oh indeed I got confused with the packet acknowledgements.
So the packet acknowledgements are queried from the destination chain https://github.com/informalsystems/hermes/blob/master/crates/relayer/src/chain/counterparty.rs?plain=1#L589, and then those sequence numbers are used to query the unreceived acks on the source chain https://github.com/informalsystems/hermes/blob/master/crates/relayer/src/chain/counterparty.rs?plain=1#L597.

File: crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                .query_unreceived_acknowledgments_sequences(
+                    dst_channel_id,
+                    dst_port_id,
+                    &acks_on_counterparty,
+                )
+                .await
+                .map_err(Relay::raise_error)?;
+
+            let ack_packets = src_chain
+                .query_send_packets_from_sequences(
```
Comment (soareschen) [inline crates/relayer/relayer-components/src/relay/impls/packet_clearers/ack.rs:85]: We should probably query for the acknowledgements from the destination chain.

**Review by soareschen — CHANGES_REQUESTED** [review]

File: crates/cosmos/cosmos-client-components/src/impls/queries/ack_packets.rs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        channel_id: &Counterparty::ChannelId,
+        port_id: &Counterparty::PortId,
+        counterparty_channel_id: &Chain::ChannelId,
+        counterparty_port_id: &Chain::PortId,
+        sequences: &[Chain::Sequence],
+        height: &Chain::Height,
+    ) -> Result<Vec<(Chain::OutgoingPacket, WriteAckEventOf<Chain, Counterparty>)>, Chain::Error>
+    {
+        let ack_packets = stream::iter(sequences)
+            .then(|sequence| {
```
Comment (soareschen) [inline crates/cosmos/cosmos-client-components/src/impls/queries/ack_packets.rs:28]: I think you might need [`for_each_concurrent`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.for_each_concurrent) to run the stream concurrently.

**Review by soareschen — APPROVED** [review]

File: crates/cosmos/cosmos-client-components/src/impls/queries/ack_packets.rs (hunk 11 above)
Comment (ljoss17) [reply crates/cosmos/cosmos-client-components/src/impls/queries/ack_packets.rs:28]: Should we also update the `send_packets.rs` with the concurrent call? https://github.com/informalsystems/hermes-sdk/blob/59ac2f025522dd021efee440c6639173f518bfc8/crates/cosmos/cosmos-client-components/src/impls/queries/send_packets.rs#L25

Comment (soareschen) [reply crates/cosmos/cosmos-client-components/src/impls/queries/ack_packets.rs:28]: Sounds good.

**Pull request merged**